File: android/dimina/src/main/kotlin/com/didi/dimina/core/Bridge.kt
package com.didi.dimina.core

import com.didi.dimina.Dimina
import com.didi.dimina.bean.BridgeOptions
import com.didi.dimina.bean.MergedPageConfig
import com.didi.dimina.bean.PathInfo
import com.didi.dimina.common.LogUtils
import com.didi.dimina.common.PathUtils
import com.didi.dimina.common.Utils
import com.didi.dimina.common.VersionUtils
import com.didi.dimina.engine.qjs.JSValue
import com.didi.dimina.ui.view.DiminaNativeComponentBridge
import com.didi.dimina.ui.container.DiminaActivity
import com.didi.dimina.ui.view.DiminaRenderBridge
import com.didi.dimina.ui.view.postMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Author: Doslin
 */
class Bridge(
    val options: BridgeOptions,
    var parent: DiminaActivity,
) {
    private val tag = "Bridge"
    private val id: String = "bridge_${Utils.uuid()}"
    @Volatile
    private var serviceResource: Boolean = false
    @Volatile
    private var renderResource: Boolean = false
    @Volatile
    private var resourceLoadedForwarded: Boolean = false
    @Volatile
    private var destroyed: Boolean = false
    @Volatile
    private var resourceLoadId: String? = null
    @Volatile
    private var desiredPageVisible: Boolean? = null
    @Volatile
    private var sentPageVisible: Boolean? = null

    /**
     * Bridge 初始化逻辑
     */
    // 保存回调引用以便在销毁时移除
    private val serviceInvokeHandler: (JSONObject) -> JSValue? = { msg -> messageInvoke("service", msg) }
    private val servicePublishHandler: (JSONObject) -> Unit = { msg -> messagePublish(msg) }

    fun init(addHandler: Boolean = true) {
        if (addHandler) {
            options.jscore.invoke(id, serviceInvokeHandler)
            options.jscore.publish(id, servicePublishHandler)

            // Add JavaScript interface for DiminaRenderBridge
            options.webview.addJavascriptInterface(
                DiminaRenderBridge(
                    invokeHandler = { msg -> messageInvoke("render", msg) },
                    publishHandler = { msg -> messagePublish(msg) }
                ), DiminaRenderBridge.TAG)
            options.webview.addJavascriptInterface(
                DiminaNativeComponentBridge(
                    touchHandler = { msg -> parent.dispatchNativeComponentTouch(msg, this@Bridge) }
                ), DiminaNativeComponentBridge.TAG)
        }
        // 加载模版页面。调试模式通过 URL 参数让 pageFrame 在 render 初始化前启用 vConsole。
        val vConsoleQuery = if (Dimina.getInstance().isDebugMode()) "?vconsole=1" else ""
        options.webview.loadUrl(
            "${PathUtils.WEBVIEW_JSSDK_BASE_URL}${VersionUtils.getJSVersion()}/main/pageFrame.html$vConsoleQuery"
        )
    }

    /**
     * Start the bridge by loading resources in both render thread and logic thread.
     */
    fun start(visible: Boolean? = null) {
        val currentResourceLoadId = synchronized(this) {
            destroyed = false
            resourceLoadId = Utils.uuid()
            if (visible != null) {
                desiredPageVisible = visible
            } else if (desiredPageVisible == null) {
                desiredPageVisible = true
            }
            resourceLoadId!!
        }

        // 通知渲染线程加载资源
        options.webview.postMessage(
            "loadResource",
            mapOf(
                "bridgeId" to id,
                "resourceLoadId" to currentResourceLoadId,
                "appId" to options.appId,
                "pagePath" to options.pathInfo.pagePath,
                "root" to options.root,
                "baseUrl" to PathUtils.WEBVIEW_JSAPP_BASE_URL
            )
        )

        // 使用协程在后台线程执行耗时的JS操作
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val jsFilePath = File(
                    parent.filesDir,
                    "jsapp/${options.appId}/${options.root}/logic.js"
                ).absolutePath

                // evaluateFromFile 会自动处理重复加载的问题
                options.jscore.evaluateFromFile(jsFilePath)

                // 发送 loadResource 消息
                options.jscore.postMessage(
                    "loadResource",
                    mapOf(
                        "bridgeId" to id,
                        "resourceLoadId" to currentResourceLoadId,
                        "appId" to options.appId,
                        "pagePath" to options.pathInfo.pagePath,
                        "root" to options.root,
                        "baseUrl" to PathUtils.WEBVIEW_JSAPP_BASE_URL
                    )
                )

                LogUtils.d(tag, "Bridge started and resources loaded in background thread")
            } catch (e: Exception) {
                LogUtils.e(tag, "Error starting bridge: ${e.message}")
            }
        }
    }

    /**
     * 消息处理
     */
    private fun messageInvoke(source: String, msg: JSONObject): JSValue? {
        if (destroyed) {
            return null
        }

        val body = msg.getJSONObject("body")
        val bridgeId = body.optString("bridgeId")

        if (bridgeId.isNotEmpty() && bridgeId != id) {
            return JSValue.createUndefined()
        }
        val incomingResourceLoadId = body.optString("resourceLoadId")
        if (incomingResourceLoadId.isNotEmpty() && incomingResourceLoadId != resourceLoadId) {
            return JSValue.createUndefined()
        }

        val type = msg.getString("type")
        val target = msg.getString("target")
        LogUtils.d(tag, "[Container] receive msg from $source: $msg")

        // Create transMsg object
        val transMsg = JSONObject().apply {
            put("type", type)
            put("body", JSONObject().apply {
                put("bridgeId", id)
                put("pagePath", options.pathInfo.pagePath)
                put("scene", options.scene)
                put("query", options.pathInfo.query)
                // Copy all properties from original body
                for (key in body.keys()) {
                    put(key, body.get(key))
                }
            })
        }

        when (target) {
            "service" -> {
                when (type) {
                    "serviceResourceLoaded" -> {
                        if (markResourceLoaded(service = true)) {
                            transMsg.put("type", "resourceLoaded")
                        } else {
                            return null
                        }
                    }

                    "renderResourceLoaded" -> {
                        if (markResourceLoaded(service = false)) {
                            transMsg.put("type", "resourceLoaded")
                        } else {
                            return null
                        }
                    }

                    "renderResourceLoadFailed" -> {
                        synchronized(this) {
                            renderResource = false
                            resourceLoadedForwarded = false
                        }
                    }
                }
                forwardToService(transMsg)
            }

            "container" -> {
                if (type == "invokeAPI") {
                    // 调用容器侧 API
                    return handleApiInvocation(body)
                } else if (type == "domReady") {
                    // 隐藏 Loading
                    parent.onDomReady()
                }
            }
        }
        return null
    }

    /**
     * 消息中转
     */
    private fun messagePublish(msg: JSONObject) {
        if (destroyed) {
            return
        }
        val body = msg.getJSONObject("body")
        val bridgeId = body.optString("bridgeId")
        if (bridgeId.isNotEmpty() && bridgeId != id) {
            return
        }

        val target = msg.getString("target")
        if (target == "service") {
            //  转发到逻辑线程
            options.jscore.postMessage(msg.toString())
        } else if (target == "render") {
            // 转发到渲染线程
            options.webview.postMessage(msg.toString())
        }


    }

    private fun isResourceLoaded(): Boolean {
        return serviceResource && renderResource
    }

    @Synchronized
    private fun markResourceLoaded(service: Boolean): Boolean {
        if (destroyed) {
            return false
        }
        if (service) {
            serviceResource = true
        } else {
            renderResource = true
        }
        if (!isResourceLoaded() || resourceLoadedForwarded) {
            return false
        }
        resourceLoadedForwarded = true
        return true
    }

    @Synchronized
    private fun forwardToService(msg: JSONObject) {
        if (destroyed) {
            return
        }
        options.jscore.postMessage(msg.toString())
        if (msg.optString("type") == "resourceLoaded") {
            flushPageVisibility()
        }
    }

    fun appShow() {
        if (!isResourceLoaded()) {
            return
        }
        options.jscore.postMessage(type="appShow")
    }

    fun appHide() {
        if (!isResourceLoaded()) {
            return
        }
        options.jscore.postMessage(type="appHide")
    }

    fun pageShow() {
        desiredPageVisible = true
        flushPageVisibility()
    }

    fun pageHide() {
        desiredPageVisible = false
        flushPageVisibility()
    }

    @Synchronized
    private fun flushPageVisibility() {
        val visible = desiredPageVisible
        if (!isResourceLoaded() || visible == null || sentPageVisible == visible) {
            return
        }

        options.jscore.postMessage(
            if (visible) "pageShow" else "pageHide",
            mapOf("bridgeId" to id)
        )
        sentPageVisible = visible
    }

    fun destroy(keepHandler: Boolean = false) {
        parent.clearNativeComponents(this)
        val wasResourceLoaded = isResourceLoaded()
        synchronized(this) {
            destroyed = true
            serviceResource = false
            renderResource = false
            resourceLoadedForwarded = false
            resourceLoadId = null
            desiredPageVisible = null
            sentPageVisible = null
        }

        // 发送页面卸载消息
        if (wasResourceLoaded) {
            options.jscore.postMessage("pageUnload", mapOf("bridgeId" to id))
        }

        if (!keepHandler) {
            // 移除在 init() 中添加的回调
            options.jscore.removeInvoke(id, serviceInvokeHandler)
            options.jscore.removePublish(id, servicePublishHandler)
        }

        LogUtils.d(tag, "Bridge destroyed and callbacks removed")
    }

    /**
     * Handles API invocation from JavaScript
     */
    private fun handleApiInvocation(body: JSONObject): JSValue? {
        try {
            val apiName = body.getString("name")

            // Handle params that could be either a JSONObject or a string
            val params = when {
                // Case 1: params is already a JSONObject
                body.optJSONObject("params") != null -> body.getJSONObject("params")

                // Case 2: params is a string, try to parse it as JSON
                body.has("params") -> {
                    val paramsValue = body.opt("params")
                    if (paramsValue is JSONArray) {
                        JSONObject().apply { put("args", body.optJSONArray("params"))}
                    } else {
                        JSONObject().apply { put("args", body.optString("params"))}
                    }
                }

                // Case 3: params is missing or null
                else -> JSONObject()
            }

            LogUtils.d(tag, "API invocation: $apiName with params: $params")

            // Use MiniApp to handle API invocation
            return parent.runWithBridgeContext(this) {
                MiniApp.getInstance().invokeAPI(
                    appId = options.appId,
                    context = parent,
                    apiName = apiName,
                    params = params
                ) { response ->
                    // Send response back to JavaScript
                    options.jscore.postMessage(response)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            LogUtils.e(tag, "Error invoking API: ${e.message}")
            return null
        }
    }

    /**
     * 更新 Bridge 的配置信息
     * @param pathInfo 新的路径信息
     * @param root 新的根路径
     * @param configInfo 新的页面配置
     */
    fun updateOptions(
        pathInfo: PathInfo,
        root: String,
        configInfo: MergedPageConfig
    ) {


        // 更新 options 中的相关配置
        options.pathInfo = pathInfo
        options.root = root
        options.configInfo = configInfo

        LogUtils.d(tag, "Bridge options updated: pathInfo=$pathInfo, root=$root")
    }
}

File: fe/packages/container/src/core/bridge.js
import chalk from 'chalk'
import { WebView } from '@/pages/webview/webview'
import { uuid } from '@/utils/util'

export class Bridge {
	constructor(opts) {
		this.id = `bridge_${uuid()}`
		this.opts = opts
		this.webview = null
		this.jscore = opts.jscore
		this.parent = null
		this.resetStatus()
	}

	async init() {
		this.jscore.invoke(msg => this.messageInvoke('service', msg))
		this.jscore.publish(msg => this.messagePublish(msg))

		this.webview = await this.createWebview()
		this.webview.invoke(msg => this.messageInvoke('render', msg))
		this.webview.publish(msg => this.messagePublish(msg))
	}

	/**
	 * 消息中转
	 * @param {*} msg
	 */
	messagePublish(msg) {
		if (this.destroyed) {
			return
		}
		if (typeof msg === 'string') {
			msg = JSON.parse(msg)
		}
		const { body, target } = msg

		// 按 id 过滤不同的 bridge 事件
		if (body.bridgeId && body.bridgeId !== this.id) {
			return
		}

		if (target === 'service') {
			this.jscore.postMessage(msg)
		}
		else if (target === 'render') {
			this.webview.postMessage(msg)
		}
	}

	/**
	 * 消息处理
	 * @param {*} source
	 * @param {*} msg
	 */
	messageInvoke(source, msg) {
		if (this.destroyed) {
			return
		}
		// 如果浏览器开启了移动设备模式，被误识别为 android/ios，需要手动转换字符串对象
		if (typeof msg === 'string') {
			msg = JSON.parse(msg)
		}
		const { type, body, target } = msg

		// 按 id 过滤不同的 bridge 事件
		if (body.bridgeId && body.bridgeId !== this.id) {
			return
		}
		if (body.resourceLoadId && body.resourceLoadId !== this.resourceLoadId) {
			return
		}

		console.log(chalk.green(`[container] receive msg from ${source}: `), msg)

		const transMsg = {
			type,
			body: {
				bridgeId: this.id,
				pagePath: this.opts.pagePath,
				scene: this.opts.scene,
				query: this.opts.query,
				...body,
			},
		}

		if (target === 'service') {
			if (type === 'serviceResourceLoaded') {
				this.serviceResource = true
				if (this.isResourceLoaded() && !this.resourceLoadedForwarded) {
					this.resourceLoadedForwarded = true
					transMsg.type = 'resourceLoaded'
				}
				else {
					return
				}
			}
			else if (type === 'renderResourceLoaded') {
				this.renderResource = true
				if (this.isResourceLoaded() && !this.resourceLoadedForwarded) {
					this.resourceLoadedForwarded = true
					transMsg.type = 'resourceLoaded'
				}
				else {
					return
				}
			}
			else if (type === 'renderResourceLoadFailed') {
				this.renderResource = false
				this.resourceLoadedForwarded = false
				transMsg.type = 'resourceLoadFailed'
			}
			this.jscore.postMessage(transMsg)
			if (transMsg.type === 'resourceLoaded') {
				this.#flushPageVisibility()
			}
		}
		else if (target === 'container') {
			if (type === 'invokeAPI') {
				const { name, params } = body
				// parent 是 miniApp 对象
				this.parent.invokeApi(name, params)
			}
		}
	}

	/**
	 * 启动资源加载
	 */
	start(options = {}) {
		this.resourceLoadId = uuid()
		if (Object.prototype.hasOwnProperty.call(options, 'visible')) {
			this.desiredPageVisible = options.visible
		}
		else if (this.desiredPageVisible === null) {
			this.desiredPageVisible = true
		}

		// 通知渲染线程加载资源
		this.webview.postMessage({
			type: 'loadResource',
			body: {
				bridgeId: this.id,
				resourceLoadId: this.resourceLoadId,
				appId: this.opts.appId,
				pagePath: this.opts.pagePath,
				root: this.opts.root,
				baseUrl: import.meta.env.BASE_URL,
			},
		})

		// 通知逻辑线程加载资源
		this.jscore.postMessage({
			type: 'loadResource',
			body: {
				bridgeId: this.id,
				resourceLoadId: this.resourceLoadId,
				appId: this.opts.appId,
				pagePath: this.opts.pagePath,
				scene: this.opts.scene,
				query: this.opts.query,
				root: this.opts.root,
				baseUrl: import.meta.env.BASE_URL,
				hostEnv: this.parent.getHostEnvSnapshot(),
			},
		})

		if (this.opts.isRoot) {
			this.jscore.postMessage({
				type: 'onUpdateStatusChange',
				body: {
					bridgeId: this.id,
					event: 'noupdate',
				},
			})
		}
	}

	resetStatus() {
		this.destroyed = false
		this.serviceResource = false
		this.renderResource = false
		this.resourceLoadedForwarded = false
		this.resourceLoadId = null
		this.desiredPageVisible = null
		this.sentPageVisible = null
	}

	createWebview() {
		return new Promise((resolve) => {
			const webview = new WebView({
				configInfo: this.opts.configInfo,
				isRoot: this.opts.isRoot,
			})

			webview.parent = this
			webview.init(() => {
				resolve(webview)
			})
			if (!this.opts.isRoot) {
				webview.el.classList.add('dimina-native-view--before-enter')
			}
			this.parent.webviewsContainer.appendChild(webview.el)
		})
	}

	/**
	 * 双线程资源是否已经初始化完成
	 */
	isResourceLoaded() {
		return this.serviceResource && this.renderResource
	}

	appShow() {
		if (!this.isResourceLoaded()) {
			return
		}
		this.jscore.postMessage({
			type: 'appShow',
			body: {},
		})
	}

	appHide() {
		if (!this.isResourceLoaded()) {
			return
		}
		this.jscore.postMessage({
			type: 'appHide',
			body: {},
		})
	}

	pageShow() {
		this.desiredPageVisible = true
		this.#flushPageVisibility()
	}

	pageHide() {
		this.desiredPageVisible = false
		this.#flushPageVisibility()
	}

	#flushPageVisibility() {
		if (
			!this.isResourceLoaded()
			|| this.desiredPageVisible === null
			|| this.sentPageVisible === this.desiredPageVisible
		) {
			return
		}

		this.jscore.postMessage({
			type: this.desiredPageVisible ? 'pageShow' : 'pageHide',
			body: {
				bridgeId: this.id,
			},
		})
		this.sentPageVisible = this.desiredPageVisible
	}

	destroy() {
		const wasResourceLoaded = this.isResourceLoaded()
		this.destroyed = true
		this.serviceResource = false
		this.renderResource = false
		this.resourceLoadedForwarded = false
		this.resourceLoadId = null
		this.desiredPageVisible = null
		this.sentPageVisible = null
		if (wasResourceLoaded) {
			this.jscore.postMessage({
				type: 'pageUnload',
				body: {
					bridgeId: this.id,
				},
			})
		}
	}
}
